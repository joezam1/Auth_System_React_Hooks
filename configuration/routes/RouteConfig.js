const routeConfiguration = Object.freeze({
    home:'/',
    authLoginPath:'/auth/login',
    authRegisterPath:'/auth/register',
    authLogoutPath:'/auth/logout',
    privateCustomerDashboard:'/customerDashboard',
    privateCustomerOrders:'/customerOrders',
});

export default routeConfiguration;