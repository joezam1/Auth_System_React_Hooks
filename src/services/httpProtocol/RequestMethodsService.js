'use strict'

import HttpRequest from './HttpRequests.js';
import HttpRequestMethods from '../../library/enumerations/HttpRequestMethods.js';



//Test:DONE
let getMethod = function (url, responseCallback, selectedHeaders = null) {
    let options = {
        method: HttpRequestMethods[HttpRequestMethods.GET],
        mode:'cors',
        headers: getHeaders(selectedHeaders)
    }

    HttpRequest.fetchMethod(url, options, responseCallback);
}

//Test:DONE
let postMethod = function (url, payload, responseCallback, selectedHeaders = null) {
    let jsonPayload = JSON.stringify(payload);

    let options = {
        method: HttpRequestMethods[HttpRequestMethods.POST],
        mode:'cors',
        headers: getHeaders(selectedHeaders),
        body: jsonPayload
    }

    HttpRequest.fetchMethod(url, options, responseCallback);
}

//Test: DONE
let deleteMethod = function (url, payload, responseCallback, selectedHeaders = null) {
    let jsonPayload = JSON.stringify(payload);

    let options = {
        method: HttpRequestMethods[HttpRequestMethods.DELETE],
        mode:'cors',
        headers: getHeaders(selectedHeaders),
        body: jsonPayload
    }

    HttpRequest.fetchMethod(url, options, responseCallback);
}

const service = Object.freeze({
    getMethod : getMethod,
    postMethod : postMethod,
    deleteMethod : deleteMethod
});

export default service;



//#REGION Private Methods

function getHeaders(selectedHeadersObj) {
    let defaultHeadersObj = {
        'Content-Type':'application/json'
    }
    let authHeaders = null ; //TODO: getAuthenticationHeaders();
    let headers = Object.assign({},defaultHeadersObj,authHeaders,selectedHeadersObj)
    return headers;
}

//#ENDREGION Private Methods