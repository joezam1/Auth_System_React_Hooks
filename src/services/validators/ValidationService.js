import ObjectModelInspector from './objectModelInspector.js';
import Helpers from '../../library/common/Helpers.js';
import InputCommonInspector from './InputCommonInspector.js';


//Test: DONE
let resolveUserFormValidation = function(userModel){
    let reportInputLength = ObjectModelInspector.inspectInputLength(userModel);
    let reportInputType = ObjectModelInspector.inspectInputType(userModel);
    let reportInputValue = ObjectModelInspector.inspectInputValue(userModel);

    let errorsReport = Object.assign({},reportInputType,reportInputLength,reportInputValue);
    return errorsReport;
}
//Test:DONE
let buildErrorMessagesReport = function(errorsReportForTargetObject, targetObject){
    let errorsSectionObjReport = {};
    let errors = errorsReportForTargetObject;
    for(let targetObjKey in targetObject){
        if(targetObject.hasOwnProperty(targetObjKey)){
            let errorSection = { objKey : '', objValue: '' };
            for(let errorKey in errors){
                if(errors.hasOwnProperty(errorKey)){
                    if(InputCommonInspector.errorKeyAndTargetKeyAreEqual(errorKey, targetObjKey)){
                        errorSection.objKey = targetObjKey + 'AllErrors';
                        let value = errors[errorKey];
                        errorSection.objValue += Helpers.getmessageFormatForDisplay(value) + Helpers.getHtmlBreakSeparator(value)
                    }
                }
            }
            if(!InputCommonInspector.stringIsNullOrEmpty(errorSection.objValue)){
                errorsSectionObjReport[errorSection.objKey] = errorSection.objValue;
            }
        }
    }
    return errorsSectionObjReport;
}

let validationManagerService = {
    resolveUserFormValidation : resolveUserFormValidation,
    buildErrorMessagesReport : buildErrorMessagesReport
}

export default validationManagerService;

//#REGION Private Functions

//#ENDREGION Private Functions
