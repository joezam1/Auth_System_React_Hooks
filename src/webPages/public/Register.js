import React, { useState } from 'react';
import { Link , Navigate } from 'react-router-dom';
import User from '../../viewModels/UserRegister.js';
import UserRoles from '../../library/enumerations/UserRoles.js';
import RequestMethodsService from '../../services/httpProtocol/RequestMethodsService.js';
import EnvConfig from '../../../configuration/environment/EnvConfig.js';
import ServerConfig from '../../../configuration/server/ServerConfig.js';
import RouteConfig from '../../../configuration/routes/RouteConfig.js';
import httpResponseStatus from '../../library/enumerations/HttpResponseStatus.js';
import InputCommonInspector from '../../services/validators/InputCommonInspector.js';
import ValidationManager from '../../services/validators/ValidationService.js';
import NotificationService from '../../services/notifications/NotificationService.js';

export default function Register(){

    const [ notificationInfo, setNotification ] = useState('');
    const [ isRegistered, setRegisterStatus] = useState(false);
    const [ firstNameErrors, setFirstNameErrors ] = useState('');
    const [ lastNameErrors, setLastNameErrors ]= useState('');
    const [ usernameErrors, setUsernameErrors ] = useState('');
    const [ emailErrors, setEmailErrors ] = useState('');
    const [ passwordErrors, setPasswordErrors ] = useState('');
    const [ confirmPasswordErrors, setConfirmPasswordErrors ] = useState('');
    let userInfo = null;

    let registerUserCallback = (response) =>{
        console.log('RegisterUserCallback-response', response);
        switch(response.status){
            case httpResponseStatus._201created:
                setNotificationPromise(NotificationService.registrationSuccess)
                .then(setRegisterStatusPromise(true));
            break;

            case httpResponseStatus._401unauthorized:
                let responseObj = (typeof response.result === 'object')
                if(responseObj){
                    setNotification(NotificationService.errorsInForm);
                    console.log('userInfo', userInfo);
                    let errorMessagesReport = ValidationManager.buildErrorMessagesReport(response.result, userInfo);
                    setAllErrorMessages(errorMessagesReport);
                    break;
                }
                setNotification(response.result);
            break;

            case httpResponseStatus._400badRequest:
                let messageRegistrationFailed = NotificationService.registrationFailed + response.result;
                setNotification(messageRegistrationFailed);
            break;

            default:
                let messageRegistrationNonProcessable = NotificationService.registrationNonProcessable + response.result;
                setNotification(messageRegistrationNonProcessable);
            break;
        }
    }

    function setNotificationPromise(message){
        let promise = new Promise(function (resolve, reject){
            resolve(setNotification(message));
        });
        return promise;
    }

    function setRegisterStatusPromise(status){
        let promise = new Promise(function(resolve, reject){
            //NOTE: Wait 5 seconds to display the SUCCESS registering the account,
            //Then redirect the user to the Login Page
            setTimeout(function(){resolve(setRegisterStatus(status)); }, 5000);
        })
        return promise;
    }

    function processUserRegistration(){
        setNotification('');
        var form = document.getElementById('registerCustomerForm');
        if(form !==null) {
            console.log('processUserRegistration-form', form);
            let dataModel = {
                firstName :(!InputCommonInspector.objectIsValid(form[0]) ) ? form[0].value : '',
                lastName : (!InputCommonInspector.objectIsValid(form[1]) ) ? form[1].value : '',
                username : (!InputCommonInspector.objectIsValid(form[2]) ) ? form[2].value : '',
                email : (!InputCommonInspector.objectIsValid(form[3]) ) ? form[3].value : '',
                password : (!InputCommonInspector.objectIsValid(form[4]) ) ? form[4].value: '',
                confirmPassword : (!InputCommonInspector.objectIsValid(form[5]) ) ? form[5].value : '',
                userRole : UserRoles.Customer
            };
            var user = new User(dataModel);
            console.log('processUserRegistration-user', user);
            userInfo = user;
            if( !inputsAreValid(user) ){ return; }
            var registerUrl = EnvConfig.PROTOCOL +'://' + EnvConfig.TARGET_URL + ServerConfig.apiUsersRegisterPathPost;
            RequestMethodsService.postMethod(registerUrl, dataModel, registerUserCallback);
        }
    }

    function inputsAreValid(objModel){
        let errorsReport = ValidationManager.resolveUserFormValidation(objModel);
        if(!InputCommonInspector.objectIsNullOrEmpty(errorsReport)){
            setNotification(messageErrorsInForm);
            let errorMessagesReport = ValidationManager.buildErrorMessagesReport(errorsReport, objModel);
            setAllErrorMessages(errorMessagesReport);
            return false;
        }
        return true;
    }

    function setAllErrorMessages(errorsReport){
        setFirstNameErrors(errorsReport.firstNameAllErrors);
        setLastNameErrors(errorsReport.lastNameAllErrors);
        setUsernameErrors(errorsReport.usernameAllErrors);
        setEmailErrors(errorsReport.emailAllErrors);
        setPasswordErrors(errorsReport.passwordAllErrors);
        setConfirmPasswordErrors(errorsReport.confirmPasswordAllErrors);
    }

    if(isRegistered){
        return <Navigate to = {RouteConfig.authLoginPath} />
    }

    return(
     <div className="register-container">
            <h3>Register</h3>
            <span><Link to={ RouteConfig.home } data-testid="link-register-home-id"> Home </Link></span>
            <br/>
            <span><Link to={ RouteConfig.authLoginPath } data-testid="link-register-login-id"> Login </Link></span>
            <br />
            <fieldset>
                <form action="/api/register" method="post" id="registerCustomerForm" className="form">
                    <input type="text" name="firstName" placeholder="Insert your first Name" className="row" onFocus={()=>{ setFirstNameErrors('');}}/>
                    <div className='text-left'>
                        <span dangerouslySetInnerHTML={{ __html: firstNameErrors }}></span>
                    </div>
                    <input type="text" name="lastName" placeholder="Insert your last Name" className="row" onFocus={()=>{setLastNameErrors('');}}/>
                    <div className='text-left'>
                        <span dangerouslySetInnerHTML={{ __html: lastNameErrors }}></span>
                    </div>
                    <input type="text" name="username" placeholder="Insert your username" className="row" onFocus={()=>{setUsernameErrors('');}}/>
                    <div className='text-left'>
                        <span dangerouslySetInnerHTML={{ __html: usernameErrors }}></span>
                    </div>
                    <input type="email" name="enail" placeholder="Insert your email" className="row" onFocus={()=>{setEmailErrors('');}}/>
                    <div className='text-left'>
                        <span dangerouslySetInnerHTML={{ __html: emailErrors }}></span>
                    </div>
                    <input type="password" name="password" placeholder="insert your password" className="row" onFocus={()=>{setPasswordErrors('');}}/>
                    <div className='text-left'>
                        <span dangerouslySetInnerHTML={{ __html: passwordErrors }}></span>
                    </div>
                    <input type="password" name="confirmPassword" placeholder="Confirm your password" className="row" onFocus={()=>{setConfirmPasswordErrors('');}}/>
                    <div className='text-left'>
                        <span dangerouslySetInnerHTML={{ __html: confirmPasswordErrors }}></span>
                    </div>

                    <button type="button" name="button" onClick={() => { processUserRegistration(); } } value="submit" className="row" data-testid="register-customer-form-id">Submit</button>
                </form>
            </fieldset>
        <p>{notificationInfo}</p>
    </div>);
}