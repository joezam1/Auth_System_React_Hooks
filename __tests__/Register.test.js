import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import { render, cleanup, fireEvent } from '@testing-library/react';
import Register from '../src/webPages/public/Register.js';
import RouteConfig from '../configuration/routes/RouteConfig.js';
import RequestMethodsService from '../src/services/httpProtocol/RequestMethodsService.js';
import ValidationService from '../src/services/validators/ValidationService.js';

jest.mock('../src/services/httpProtocol/RequestMethodsService.js');
jest.mock('../src/services/validators/validationService.js');

describe('File: Register.js', function(){
    afterEach(cleanup);
    afterAll(()=>{
        jest.resetAllMocks();
    });
    let container;
    beforeEach(()=>{
        container = render( <MemoryRouter> <Register/> </MemoryRouter>);
    });
    describe('Component: Register', function(){
        test('It can take a snapshot', function(){
            expect(container.asFragment(<Register/>)).toMatchSnapshot();
        });

        test('Home Link Redirects Correctly', function(){
            const link = container.getByText( 'Home' );
            const link1 = container.getByTestId ('link-register-home-id');
            fireEvent.click(link);
            var initialPath = window.location.pathname;
            var pathname = link.pathname;
            expect(pathname).toBe(RouteConfig.home);
            expect(link).toBe(link1);
        });

        test('Login Link Redirects Correctly', function(){
            const link = container.getByText( 'Login' );
            const link1 = container.getByTestId ('link-register-login-id');
            fireEvent.click(link);
            var initialPath = window.location.pathname;
            var pathname = link.pathname;
            expect(pathname).toBe(RouteConfig.authLoginPath);
            expect(link).toBe(link1);
        });

        test('submit button Triggers form', function(){
            //Arrange
            const submitButton = container.getByTestId('register-customer-form-id');
            const resultMock = 'OK';
            const resultValidationMock = {};
            ValidationService.resolveUserFormValidation = jest.fn().mockReturnValueOnce(resultValidationMock);
            RequestMethodsService.postMethod = jest.fn().mockReturnValueOnce(resultMock);
            //Act

            fireEvent.click(submitButton);
            //Assert
            expect(RequestMethodsService.postMethod).toHaveBeenCalledTimes(1);
            expect(ValidationService.resolveUserFormValidation).toHaveBeenCalledTimes(1);
        });
    });
});