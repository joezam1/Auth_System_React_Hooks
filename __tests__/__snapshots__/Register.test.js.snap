// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File: Register.js Component: Register It can take a snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="register-container"
  >
    <h3>
      Register
    </h3>
    <span>
      <a
        data-testid="link-register-home-id"
        href="/"
      >
         Home 
      </a>
    </span>
    <br />
    <span>
      <a
        data-testid="link-register-login-id"
        href="/auth/login"
      >
         Login 
      </a>
    </span>
    <br />
    <fieldset>
      <form
        action="/api/register"
        class="form"
        id="registerCustomerForm"
        method="post"
      >
        <input
          class="row"
          name="firstName"
          placeholder="Insert your first Name"
          type="text"
        />
        <div
          class="text-left"
        >
          <span />
        </div>
        <input
          class="row"
          name="lastName"
          placeholder="Insert your last Name"
          type="text"
        />
        <div
          class="text-left"
        >
          <span />
        </div>
        <input
          class="row"
          name="username"
          placeholder="Insert your username"
          type="text"
        />
        <div
          class="text-left"
        >
          <span />
        </div>
        <input
          class="row"
          name="enail"
          placeholder="Insert your email"
          type="email"
        />
        <div
          class="text-left"
        >
          <span />
        </div>
        <input
          class="row"
          name="password"
          placeholder="insert your password"
          type="password"
        />
        <div
          class="text-left"
        >
          <span />
        </div>
        <input
          class="row"
          name="confirmPassword"
          placeholder="Confirm your password"
          type="password"
        />
        <div
          class="text-left"
        >
          <span />
        </div>
        <button
          class="row"
          data-testid="register-customer-form-id"
          name="button"
          type="button"
          value="submit"
        >
          Submit
        </button>
      </form>
    </fieldset>
    <p />
  </div>
   
</DocumentFragment>
`;
